File: android/app/src/main/java/com/lightningkite/kwifttemplate/vg/BleScanDemoVG.shared.kt
package com.lightningkite.kwifttemplate.vg

import android.widget.*
import android.view.*
import com.lightningkite.kwift.*
import com.lightningkite.kwift.bluetooth.Ble
import com.lightningkite.kwift.bluetooth.BleDeviceInfo
import com.lightningkite.kwift.lifecycle.*
import com.lightningkite.kwift.views.*
import com.lightningkite.kwift.observables.*
import com.lightningkite.kwift.observables.binding.*
import com.lightningkite.kwifttemplate.R
import com.lightningkite.kwifttemplate.layouts.*

@Suppress("NAME_SHADOWING")
class BleScanDemoVG(
    val stack: ObservableStack<ViewGenerator>
) : ViewGenerator() {

    override val title: String get() = "BleScanDemo"

    val devices: MutableObservableProperty<Map<String, BleDeviceInfo>> = StandardObservableProperty(mapOf())

    override fun generate(dependency: ViewDependency): View {
        val xml = BleScanDemoXml()
        val view = xml.setup(dependency)

        xml.button.onClick(captureWeak(xml.button){ button ->
            val closeable = Ble.scan(
                viewDependency = dependency,
                withServices = listOf(),
                intensity = .5f,
                onDeviceFound = { device ->
                    println("Found device $device")
                    this.devices.value = this.devices.value + Pair(device.id, device)
                }
            )
            (button.lifecycle and appInForeground).closeWhenOff(closeable)
        })

        xml.host.onClick {
            this.stack.push(BleHostDemoVG())
        }

        xml.items.bind(
            data = devices.map { it -> it.values.sortedByDescending { it -> it.rssi } },
            defaultValue = BleDeviceInfo("", "", 0),
            makeView = label@ { obs ->
                val cellXml = ComponentBleDeviceXml()
                val cellView = cellXml.setup(dependency)
                cellXml.deviceName.bindString(obs.map { it -> it.name })
                cellXml.deviceId.bindString(obs.map { it -> it.id })
                cellXml.rssi.bindString(obs.map { it -> it.rssi.toString() })
                cellXml.connect.onClick {  }
                return@label cellView
            }
        )

        return view
    }

}

File: ios/Kwift Template/com/lightningkite/kwifttemplate/vg/BleScanDemoVG.shared.swift
//Package: com.lightningkite.kwifttemplate.vg
//Converted using Kwift2

import Foundation
import Kwift
import RxSwift
import RxRelay
import KwiftMaps



public class BleScanDemoVG: ViewGenerator {
    
    public var stack: ObservableStack<ViewGenerator>
    
    override public var title: String {
        get {
            return "BleScanDemo"
        }
    }
    public var devices: MutableObservableProperty<Dictionary<String, BleDeviceInfo>>
    
    override public func generate(dependency: ViewDependency) -> View {
        var xml = BleScanDemoXml()
        var view = xml.setup(dependency)
        xml.button.onClick(captureWeak(xml.button) { (button) in 
            var closeable = Ble.scan(viewDependency: dependency, withServices: [], intensity: Float(0.5), onDeviceFound: { (device) in 
                print("Found device \(device)")
                self.devices.value = self.devices.value + (device.id, device)
            })
            ( button.lifecycle.and(appInForeground) ).closeWhenOff(closeable)
        })
        xml.host.onClick{ () in 
            self.stack.push(BleHostDemoVG())
        }
        xml.items.bind(data: devices.map{ (it) in 
            it.values.sortedByDescending{ (it) in 
                it.rssi
            }
        }, defaultValue: BleDeviceInfo("", "", 0), makeView: { (obs) in 
            var cellXml = ComponentBleDeviceXml()
            var cellView = cellXml.setup(dependency)
            cellXml.deviceName.bindString(obs.map{ (it) in 
                it.name
            })
            cellXml.deviceId.bindString(obs.map{ (it) in 
                it.id
            })
            cellXml.rssi.bindString(obs.map{ (it) in 
                it.rssi.toString()
            })
            cellXml.connect.onClick{ () in 
            }
            return cellView
        })
        return view
    }
    override public func generate(_ dependency: ViewDependency) -> View {
        return generate(dependency: dependency)
    }
    
    public init(stack: ObservableStack<ViewGenerator>) {
        self.stack = stack
        let devices: MutableObservableProperty<Dictionary<String, BleDeviceInfo>> = StandardObservableProperty([:])
        self.devices = devices
        super.init()
    }
    convenience public init(_ stack: ObservableStack<ViewGenerator>) {
        self.init(stack: stack)
    }
}
 
